### PR TITLE
ARROW-4330: [C++] More robust discovery of pthreads

### PR DESCRIPTION
**Pull request by wesm**
I found when using a non-system gcc (e.g. the `gcc_linux-64` and `gxx_linux-64` gcc7 conda packages) that CMake is not able to discover pthreads. The conda-forge community regards this as a bug in CMake, but there is another better way to do this using the built-in `FindThreads.cmake`.

There's a chance this might also resolve the missing thread symbol in the Gandiva windows build, but I haven't tried it yet

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -839,50 +839,40 @@ if (ARROW_BUILD_BENCHMARKS)
...
-  list(APPEND
-    ARROW_SHARED_INSTALL_INTERFACE_LIBS
-    pthread)
-  list(APPEND
-    ARROW_STATIC_LINK_LIBS
-    pthreadshared)
+if (THREADS_FOUND)
+  set(ARROW_SYSTEM_LINK_LIBS
+    ${ARROW_SYSTEM_LINK_LIBS}
+    Threads::Threads)
```
Comment (kou) [inline cpp/CMakeLists.txt]: `list(APPEND ARROW_SYSTEM_LINK_LIBS Threads::Threads)` is better.

**Review by kou — APPROVED** [review]
+1 with a small fix.

**Review by xhochy — APPROVED** [review]
+1, LGTM

I have restarted the OSX build, failure seems unrelated.

**Comment by xhochy** [comment]
OSX passed, R will be fixed with https://github.com/apache/arrow/pull/3500

**Pull request closed without merging**